packages/flutter_tools/lib/src/build_system/targets/macos.dart

```diff
@@ -189,19 +191,30 @@ class DebugBundleFlutterAssets extends Target {
   @override
   Future<void> build(List<File> inputFiles, Environment environment) async {
     final FlutterProject flutterProject = FlutterProject.fromDirectory(environment.projectDir);
-    final Directory outputDirectory = flutterProject.macos
-        .ephemeralDirectory.childDirectory('App.framework');
-    if (!outputDirectory.existsSync()) {
-      throw Exception('App.framework must exist to bundle assets.');
-    }
+    final Directory frameworkRootDirectory = flutterProject.macos
+        .ephemeralDirectory
+        .childDirectory('App.framework');
+    final Directory outputDirectory = frameworkRootDirectory
+        .childDirectory('Versions')
+        .childDirectory('A')
+        ..createSync(recursive: true);
+
+    // Copy App into framework directory.
+    environment.buildDir
+      .childDirectory('App.framework')
+      .childFile('App')
```

Comment:
This binary should also be a symlink from frameworkRootDirectory, not a copy.

Just to look at the anatomy of an embedded Xcode framework (because if they aren't getting it right then no one is) at Xcode.app/Contents/Frameworks/DFRSupportKit.framework:
```
lrwxr-xr-x  1  primarygroup     30 Jul 29 14:36 ./DFRSupportKit -> Versions/Current/DFRSupportKit
lrwxr-xr-x  1  primarygroup     26 Jul 29 14:36 ./Resources -> Versions/Current/Resources
drwxr-xr-x  4  primarygroup    128 Jul 29 14:51 ./Versions
drwxr-xr-x  5  primarygroup    160 Jul 26 10:01 ./Versions/A
-rwxr-xr-x  1  primarygroup  22704 Jul 26 10:01 ./Versions/A/DFRSupportKit
drwxr-xr-x  4  primarygroup    128 Jul 25 02:18 ./Versions/A/Resources
-rw-r--r--  1  primarygroup    789 Jul 25 02:18 ./Versions/A/Resources/Info.plist
-rw-r--r--  1  primarygroup    521 Jul 25 02:18 ./Versions/A/Resources/version.plist
drwxr-xr-x  3  primarygroup     96 Jul 25 02:18 ./Versions/A/_CodeSignature
-rw-r--r--  1  primarygroup   2674 Jul 26 10:01 ./Versions/A/_CodeSignature/CodeResources
lrwxr-xr-x  1  primarygroup      1 Jul 29 14:36 ./Versions/Current -> A
```
The only non-symlink `-type file` files are in `Versions/A/` but the symlinks all link `Versions/Current`
```
-rwxr-xr-x  1  primarygroup  22704 Jul 26 10:01 ./Versions/A/DFRSupportKit
-rw-r--r--  1  primarygroup    789 Jul 25 02:18 ./Versions/A/Resources/Info.plist
-rw-r--r--  1  primarygroup    521 Jul 25 02:18 ./Versions/A/Resources/version.plist
-rw-r--r--  1  primarygroup   2674 Jul 26 10:01 ./Versions/A/_CodeSignature/CodeResources
```

Reply:
The binary is located under versions/A. The copy in build_dir is in a build cache under .dart_tool/flutter_build/hashed-values.